src/algebra/direct_sum_graded.lean

```diff
@@ -543,6 +545,108 @@ end comm_ring
 
 end grade_zero
 
+section to_semiring
+
+variables {R : Type*} [Π i, add_comm_monoid (A i)] [add_monoid ι] [gmonoid A] [semiring R]
+variables {A}
+
+/-- If two ring homomorphisms from `⨁ i, A i` are equal on each `of A i y`,
+then they are equal.
+
+See note [partially-applied ext lemmas]. -/
+@[ext]
+lemma ring_hom_ext' (F G : (⨁ i, A i) →+* R)
+  (h : ∀ i, (F : (⨁ i, A i) →+ R).comp (of _ i) = (G : (⨁ i, A i) →+ R).comp (of _ i)) : F = G :=
+ring_hom.coe_add_monoid_hom_injective $ direct_sum.add_hom_ext' h
+
+section to_move
+
+lemma _root_.add_monoid_hom.ext_iff₂
+  {R S T : Type*} [add_zero_class R] [add_zero_class S] [add_comm_monoid T] {f g : R →+ S →+ T} :
+  f = g ↔ (∀ x y, f x y = g x y) :=
+add_monoid_hom.ext_iff.trans $ forall_congr $ λ _, add_monoid_hom.ext_iff
+
+/-- An `add_monoid_hom` preserves multiplication if pre- and post- composition with
+`add_monoid_hom.mul` are equivalent. By converting the statement into an equality of
+`add_monoid_hom`s, various specialized `ext` lemmas about `→+` can then apply. -/
+lemma _root_.add_monoid_hom.map_mul_iff
+  {R : Type u_1} {S : Type u_2} [semiring R] [semiring S] (f : R →+ S) :
+  (∀ x y, f (x * y) = f x * f y) ↔
+    (f.comp_hom : _ →+ R →+ _).comp add_monoid_hom.mul =
+      ((add_monoid_hom.mul.comp f).flip.comp f).flip :=
+iff.symm add_monoid_hom.ext_iff₂
+
+end to_move
```

Comment:
Are you going to move this stuff? The PR is marked "awaiting review".

Reply:
Yes, I planned to move them after getting some feedback about whether these are the right definitions.